scipy/signal/_peak_finding_utils.pyx

```diff
@@ -74,3 +74,92 @@ def _argmaxima1d(np.float64_t[:] x not None):
 
     maxima.resize(m, refcheck=False)  # Keep only valid part of array memory.
     return maxima
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def _peak_prominences(np.float64_t[::1] x not None,
+                      np.intp_t[::1] peaks not None,
+                      np.intp_t wlen):
+    """
+    Calculate the prominence of each peak in a signal.
+
+    Parameters
+    ----------
+    x : ndarray
+        A signal with peaks.
+    peaks : ndarray
+        Indices of peaks in `x`.
+    wlen : np.intp
+        A window length in samples (see `peak_prominences`) which is rounded up
+        to the nearest odd integer. If smaller than 2 the entire signal `x` is
+        used.
+
+    Returns
+    -------
+    prominences : ndarray
+        The calculated prominences for each peak in `peaks`.
+    left_bases, right_bases : ndarray
+        The peaks' bases as indices in `x` to the left and right of each peak.
+
+    Raises
+    ------
+    ValueError
+        If an index in `peaks` doesn't point to a local maximum in `x`.
+
+    Notes
+    -----
+    This is the inner function to `peak_prominences`.
+
+    .. versionadded:: 1.1.0
+    """
+    cdef:
+        np.float64_t[::1] prominences
+        np.intp_t[::1] left_bases, right_bases
+        np.float64_t peak_height, left_min, right_min
+        np.intp_t peak_nr, peak, i_min, i_max, i
+
+    # Allocate returned objects
+    prominences = np.empty(peaks.shape[0], dtype=np.float64)
+    left_bases = np.empty(peaks.shape[0], dtype=np.intp)
+    right_bases = np.empty(peaks.shape[0], dtype=np.intp)
+
+    for peak_nr in range(peaks.shape[0]):
```

Comment:
Would it be possible to wrap this loop in a `with nogil:` statement? The `raise` statements would have to be moved outside but with a `raise_error_flag = True; break` the same functionality could easily be achieved. Cython's documentation says [it is safe to release the GIL for memoryviews](http://docs.cython.org/en/latest/src/userguide/memoryviews.html#memoryviews-and-the-gil) (see also [here](https://cython.readthedocs.io/en/latest/src/userguide/external_C_code.html#acquiring-and-releasing-the-gil)). However I'm not sure whether it is safe to release the GIL when indexing memoryviews that point to memory of valid Python objects (see input arrays).

Reply:
> I'm not sure whether it is safe to release the GIL when indexing memoryviews that point to memory of valid Python objects

I suspect this would be a very common use case for typed memoryviews, so if the Cython docs don't list a (prominent) caveat about it, I suspect it should be fine. But if you to ping the Cython devs to ask (and perhaps offer to update their docs to make a note of the answer!) that would be the safest route, unless someone here already knows.

Reply:
Good idea. I'll look into it. :+1:
For now, I won't add a `with nogil` until I'm clear on this.

Reply:
I can find several other places in SciPy's code base that show a comparable usage: e.g. [_kendal_dis](https://github.com/scipy/scipy/blob/master/scipy/stats/_stats.pyx#L84) or [_colloc](https://github.com/scipy/scipy/blob/636c216eaedf8a570e55ee93a8082d306f772816/scipy/interpolate/_bspl.pyx#L389).

@ev-br (I hope its okay to ping you), the last example was committed by you. Am I right in assuming this is safe to do?

Reply:
+1 for using nogil memoryviews that point to numpy array data. I'm not a Cython dev, but I've used that trick a bunch of times with no issues.